### PR TITLE
ARTEMIS-2457 implement ring queue

### PR DESCRIPTION
**Pull request by jbertram**
Don't merge this yet. I've still got a few tests to add for basic configuration stuff, and I also need to finish the docs. I'm putting this up now to get reviews/opinions.

### CODE REVIEW

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java [hunk 1]
```diff
@@ -49,14 +52,18 @@
...
+   public QueuePendingMessageMetrics(final Queue queue, final String name) {
       Preconditions.checkNotNull(queue);
       this.queue = queue;
+      this.name = name;
    }
 
    public void incrementMetrics(final MessageReference reference) {
       long size = getPersistentSize(reference);
       COUNT_UPDATER.incrementAndGet(this);
+      logger.debugf("%s increment messageCount to %d: %s", this, messageCount, reference);
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java]: can we wrap this in a isDebugEnabled

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java [hunk 2]
```diff
@@ -67,6 +74,7 @@ public void incrementMetrics(final MessageReference reference) {
    public void decrementMetrics(final MessageReference reference) {
       long size = -getPersistentSize(reference);
       COUNT_UPDATER.decrementAndGet(this);
+      logger.debugf("%s decrement messageCount to %d: %s", this, messageCount, reference);
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java]: can we wrap this in a isDebugEnabled

**Comment by michaelandrepearce** [comment]
In QPID both size in message count can be configured / or size in total bytes.  I cannot see how we can set the latter?

What's the reason to make this a brand new queue type entirely over just making it something configurable, i ask as it means cannot change an existing queue into a ring queue without deleting and recreating, where as having it a setting in QueueImpl like almost all others would give the dynamic ability. 

On top of the last comment, can we dynamically update the size, e.g. increase or decrease its limit, so that if incorrectly sized it can be tuned, without need of re-creating.

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/QueueQueryImpl.java [hunk 3]
```diff
@@ -95,70 +97,7 @@ public QueueQueryImpl(final boolean durable,
                          final SimpleString name,
                          final boolean exists,
                          final boolean autoCreateQueues) {
-      this(durable, temporary, consumerCount, messageCount, filterString, address, name, exists, autoCreateQueues, -1, false, false, RoutingType.MULTICAST);
```
Comment (michaelandrepearce) [inline artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/QueueQueryImpl.java]: wont this be a breaking change removing the old api's if anyone is using CORE api directly? I appreciate its hidden for JMS users.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQServer.java [hunk 4]
```diff
@@ -449,7 +449,7 @@ Queue createQueue(AddressInfo addressInfo, SimpleString queueName, SimpleString
    Queue createQueue(AddressInfo addressInfo, SimpleString queueName, SimpleString filter,
                      SimpleString user, boolean durable, boolean temporary, boolean autoCreated, Integer maxConsumers,
                      Boolean purgeOnNoConsumers, Boolean exclusive, Boolean groupRebalance, Integer groupBuckets, SimpleString groupFirstKey, Boolean lastValue, SimpleString lastValueKey, Boolean nonDestructive,
-                     Integer consumersBeforeDispatch, Long delayBeforeDispatch, Boolean autoDelete, Long autoDeleteDelay, Long autoDeleteMessageCount, boolean autoCreateAddress) throws Exception;
+                     Integer consumersBeforeDispatch, Long delayBeforeDispatch, Boolean autoDelete, Long autoDeleteDelay, Long autoDeleteMessageCount, boolean autoCreateAddress, Long ringSize) throws Exception;
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQServer.java]: Breaking api change, since last version, should have an override method.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ActiveMQServerImpl.java [hunk 5]
```diff
@@ -3272,7 +3273,8 @@ public Queue createQueue(final AddressInfo addrInfo,
                             final long autoDeleteDelay,
                             final long autoDeleteMessageCount,
                             final boolean autoCreateAddress,
-                            final boolean configurationManaged) throws Exception {
+                            final boolean configurationManaged,
+                            final long ringSize) throws Exception {
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ActiveMQServerImpl.java]: breaking api change, new param needs a new method added (old method interface honoured)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java (hunk 1 above)
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java]: can we put inside if debug enabled (as on hot path) so will be removed by jit if debug isnt on.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java (hunk 2 above)
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java]: can we put inside if debug enabled (as on hot path) so will be removed by jit if debug isnt on.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+   private void enforceRing() {
+      enforceRing(null, false);
+   }
+
+   private void enforceRing(MessageReference refToAck, boolean scheduling) {
+      if (getMessageCountForRing() >= ringSize) {
+         refToAck = refToAck == null ? messageReferences.poll() : refToAck;
+
+         if (refToAck != null) {
+            logger.infof("Preserving ringSize %d by acking message ref %s", ringSize, refToAck);
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java]: assume this is accidentally left in, would be very verbose logging at info

File: docs/user-manual/en/ring-queue.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+However, before we get to those use cases let's look at the basic configuration
+of a ring queue.
+
+## Configuration
+
+There are 2 parameters related to ring queue configuration.
+
+The `ring-size` parameter can be set directly on the `queue` element. The
+default value is `-1` (i.e. no limit).
```
Comment (michaelandrepearce) [inline docs/user-manual/en/ring-queue.md]: wont the default value come from address setting?

File: docs/user-manual/en/ring-queue.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+<address name="myRing">
+   <anycast>
+      <queue name="myRing" ring-size="3" />
+   </anycast>
+</address>
+```
+
+The `default-ring-size` is an `address-setting` which applies to queues on
+matching addresses which don't have an explicit `ring-size` set. This is
+especially useful for auto-created queues.
```
Comment (michaelandrepearce) [inline docs/user-manual/en/ring-queue.md]: if not set default is -1 at the address setting level

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/server/RingQueueTest.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+import org.apache.activemq.artemis.core.server.impl.RingQueue;
+import org.apache.activemq.artemis.core.settings.impl.AddressSettings;
+import org.apache.activemq.artemis.jms.client.ActiveMQConnectionFactory;
+import org.apache.activemq.artemis.tests.util.ActiveMQTestBase;
+import org.apache.activemq.artemis.tests.util.Wait;
+import org.apache.activemq.artemis.utils.RandomUtil;
+import org.junit.Before;
+import org.junit.Test;
+
+public class RingQueueTest extends ActiveMQTestBase {
```
Comment (michaelandrepearce) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/server/RingQueueTest.java]: Can we add a test with persistence, also a test with non-destructive ring queue (ensure that when non-destructive it sill replaces/acks)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+   }
+
+   @Override
+   public long getRingSize() {
+      return ringSize;
+   }
+
+   public long getMessageCountForRing() {
+      return (long) pendingMetrics.getMessageCount();// + (pageSubscription == null ? 0 : pageSubscription.getMessageCount());
+   }
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java]: hashcode and equals probably need implemented similar  like LVQ

**Comment by cshannon** [comment]
This is kind of nit picky but I would prefer a different name like "FixedSizeQueue" instead of "RingQueue" or something just because the name RingQueue immediately made me think of Ring/Circular buffers where memory is re-used for performance so I thought this would be something similar to LMAX Disruptor, etc.  However, obviously the goal here is to just to maintain a max size or fix sized queue by acking the oldest message.  But I'm certainly not going to -1 over the name, if everyone wants to keep RingQueue that's fine.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ActiveMQServerImpl.java (hunk 5 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ActiveMQServerImpl.java]: Hmmm...Even though this is public it isn't part of the interface so I'm inclined to leave this.

**Comment by michaelandrepearce** [comment]
I think in qpid it was ring queue, for those moving from qpid over its kind of nice to keep consistent but not strongly opinionated

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java (hunk 1 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java]: Will fix.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java (hunk 2 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java]: Will fix.

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/QueueQueryImpl.java (hunk 3 above)
Comment (jbertram) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/QueueQueryImpl.java]: Will fix.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQServer.java (hunk 4 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQServer.java]: Will fix.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java (hunk 1 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java]: Will fix.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java (hunk 2 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java]: Will fix.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java (hunk 6 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java]: Will fix.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+   private void enforceRing(MessageReference refToAck, boolean scheduling) {
+      if (getMessageCountForRing() >= ringSize) {
+         refToAck = refToAck == null ? messageReferences.poll() : refToAck;
+
+         if (refToAck != null) {
+            logger.infof("Preserving ringSize %d by acking message ref %s", ringSize, refToAck);
+            referenceHandled(refToAck);
+
+            try {
+               refToAck.acknowledge(null, AckReason.REPLACED, null);
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java]: IMO it's different enough to warrant a new method.

File: docs/user-manual/en/ring-queue.md (hunk 7 above)
Comment (jbertram) [reply docs/user-manual/en/ring-queue.md]: Will fix.

File: docs/user-manual/en/ring-queue.md (hunk 8 above)
Comment (jbertram) [reply docs/user-manual/en/ring-queue.md]: Will fix.

**Comment by jbertram** [comment]
@cshannon, @michaelandrepearce is right about the name coming from Qpid. I see your point, but for continuity with Qpid I think it makes sense to keep the current name.

**Comment by cshannon** [comment]
@jbertram - Alright that's fine to keep the name then since there is some precedence on the name coming from Qpid.

**Comment by jbertram** [comment]
> In QPID both size in message count can be configured / or size in total bytes. I cannot see how we can set the latter?

Only count is supported at this point.

> What's the reason to make this a brand new queue type entirely over just making it something configurable...

I was following the pattern established by the last-value queue implementation. I did consider just making it part of the queue implementation, but it seemed cleaner to separate out the ring code.

> ...can we dynamically update the size, e.g. increase or decrease its limit, so that if incorrectly sized it can be tuned, without need of re-creating.

There is no facility for changing the ring-size at this point. I'll look at adding that as it would be nice to have.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/server/RingQueueTest.java (hunk 9 above)
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/server/RingQueueTest.java]: What do you mean a test with persistence? The existing tests use persistent messages. Were you talking about testing the persistence of the queue binding encoding?

**Comment by jbertram** [comment]
I resolved the conflict, updated the code per the reviews, & added some additional tests. I'm still working on the docs, though so don't merge quite yet.

**Comment by jbertram** [comment]
Docs are done. This is ready to merge from my perspective.

**Comment by jbertram** [comment]
@wy96f, take a look at the code now. I believe I addressed your concerns.

**Comment by michaelandrepearce** [comment]
Is there added ability to update the size? I really would like this sorted before merge as it stands with it not being implemented inside the main queueimpl i beleive it makes it very hard to implement. Thus my original feedback / query to why have it as independent type, instead of implementing within queueimpl.

**Comment by michaelandrepearce** [comment]
Ill try make something this weekend to show what im meaning

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+   public boolean allowsReferenceCallback() {
+      return false;
+   }
+
+   private void enforceRing() {
+      enforceRing(null, false);
+   }
+
+   private void enforceRing(MessageReference refToAck, boolean scheduling) {
+      if (getMessageCountForRing() >= ringSize) {
```
Comment (wy96f) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RingQueue.java]: I think we should judge whether ref is already acked first.

E.g. for non destructive queue, suppose ring size is 1, m1 is sent and delivered to consumer c1. Then m2 is sent and delivered to consumer c2. Thus m1 is replace acked and m2 resides in "messageReferences"(message count is now 1). Now consumer c1 is closed without acking m1, m1 is rollbacked and replace acked again that would throw IllegalStateException.

**Comment by jbertram** [comment]
@michaelandrepearce, I've moved the ring semantics into `QueueImpl`.

@wy96f, I'm still looking at the issue you identified.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 13]
```diff
@@ -576,15 +612,17 @@ public QueueImpl(final long id,
...
 
    // Bindable implementation -------------------------------------------------------------------------------------
 
    @Override
    public boolean allowsReferenceCallback() {
-      // non descructive queues will reuse the same reference between multiple consumers
+      // non destructive queues will reuse the same reference between multiple consumers
       // so you cannot really use the callback from the MessageReference
-      return !nonDestructive;
+      return !isNonDestructive();
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: ? Why the chang. Its internal field and allows access by field reference

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 14]
```diff
@@ -2733,7 +2799,7 @@ private boolean deliver() {
    }
 
    protected void removeMessageReference(ConsumerHolder<? extends Consumer> holder, MessageReference ref) {
-      if (!nonDestructive) {
+      if (!isNonDestructive()) {
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: why the change here? again its internal to class, and hot path, thus was field access

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 15]
```diff
@@ -3436,6 +3503,11 @@ public void postAcknowledge(final MessageReference ref, AckReason reason) {
 
       queue.decDelivering(ref);
 
+      // this is done to tell the difference between actual acks and just a closed consumer in the non-destructive use-case
+      if (isNonDestructive() && reason == AckReason.NORMAL) {
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: field access?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 16]
```diff
@@ -1556,7 +1614,7 @@ public void acknowledge(final MessageReference ref, final ServerConsumer consume
 
    @Override
    public void acknowledge(final MessageReference ref, final AckReason reason, final ServerConsumer consumer) throws Exception {
-      if (nonDestructive && reason == AckReason.NORMAL) {
+      if (isNonDestructive() && reason == AckReason.NORMAL) {
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: was field access why the change, its hotpath and was internal field.

Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: i realise it maybe in-lined but no guarantee.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 17]
```diff
@@ -165,7 +165,7 @@
 
    private volatile boolean queueDestroyed = false;
 
-   private final PageSubscription pageSubscription;
+   protected final PageSubscription pageSubscription;
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: Is this needed?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 18]
```diff
@@ -179,17 +179,17 @@
    private final MpscUnboundedArrayQueue<MessageReference> intermediateMessageReferences = new MpscUnboundedArrayQueue<>(8192);
 
    // This is where messages are stored
-   private final PriorityLinkedList<MessageReference> messageReferences = new PriorityLinkedListImpl<>(QueueImpl.NUM_PRIORITIES, MessageReferenceImpl.getIDComparator());
+   protected final PriorityLinkedList<MessageReference> messageReferences = new PriorityLinkedListImpl<>(QueueImpl.NUM_PRIORITIES, MessageReferenceImpl.getIDComparator());
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: is this needed?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 19]
```diff
@@ -179,17 +179,17 @@
...
 
    // The quantity of pagedReferences on messageReferences priority list
    private final AtomicInteger pagedReferences = new AtomicInteger(0);
 
    // The estimate of memory being consumed by this queue. Used to calculate instances of messages to depage
    private final AtomicInteger queueMemorySize = new AtomicInteger(0);
 
-   private final QueuePendingMessageMetrics pendingMetrics = new QueuePendingMessageMetrics(this);
+   protected final QueueMessageMetrics pendingMetrics = new QueueMessageMetrics(this, "pending");
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: again is this needed, making it protected

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 20]
```diff
@@ -3933,6 +4005,37 @@ void scheduleSlowConsumerReaper(AddressSettings settings) {
       }
    }
 
+   private void enforceRing() {
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: could ringsize != -1 check be done here, will allow better escaping and inlining where not being used., e.g. avoids need to call into method below which is many lines long, and unlikely to be inlined

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 21]
```diff
@@ -881,13 +919,17 @@ public void addHead(final MessageReference ref, boolean scheduling) {
       enterCritical(CRITICAL_PATH_ADD_HEAD);
       synchronized (this) {
          try {
-            if (!scheduling && scheduledDeliveryHandler.checkAndSchedule(ref, false)) {
-               return;
-            }
+            enforceRing(ref, scheduling);
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: can a if check be wrapped around this, to avoid the method (which method is large and unlikly to be inlined) call when ringsize is disabled (-1)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/GroupFirstMessageReference.java [hunk 22]
```diff
@@ -191,6 +191,16 @@ public void handled() {
       messageReference.handled();
    }
 
+   @Override
+   public void setInDelivery(boolean deliveredDirectly) {
+
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/GroupFirstMessageReference.java]: why is it not invoking?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 23]
```diff
@@ -3460,7 +3532,7 @@ public void postAcknowledge(final MessageReference ref, AckReason reason) {
          message = null;
       }
 
-      if (message == null)
+      if (message == null || (isNonDestructive() && reason == AckReason.NORMAL))
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: by field access

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 24]
```diff
@@ -576,15 +612,17 @@ public QueueImpl(final long id,
       if (this.addressInfo != null && this.addressInfo.isPaused()) {
          this.pause(false);
       }
+
+      this.ringSize = ringSize == null ? (addressSettingsRepository != null ? addressSettingsRepository.getMatch(address.toString()).getDefaultRingSize() : -1) : ringSize;
```
Comment (michaelpearce-gain) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: doesnt seem to match the logic of all others, which is that address setting resolvement if not set is meant to occur well before create queue object (e.g. ownership of that logic is else where) to keep things standardised where looking for where things occur, and if null reaches here (not expected, but a safety is that it defaults to the constant.

Currently if it was null and addresssettingrepo is null, you will get NPE. which others dont.


e.g.

expecting something more inline with others:

example:
      this.maxConsumers = maxConsumers == null ? ActiveMQDefaultConfiguration.getDefaultMaxQueueConsumers() : maxConsumers;

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java [hunk 25]
```diff
@@ -335,6 +335,66 @@ public void testCreateAndDestroyQueue_4() throws Exception {
...
+      Assert.assertEquals(durable, queueControl.isDurable());
+      Assert.assertEquals(maxConsumers, queueControl.getMaxConsumers());
+      Assert.assertEquals(purgeOnNoConsumers, queueControl.isPurgeOnNoConsumers());
+      Assert.assertEquals(false, queueControl.isTemporary());
+      Assert.assertEquals(exclusive, queueControl.isExclusive());
+//      Assert.assertEquals(groupRebalance, queueControl.getGroupRebalance());
+//      Assert.assertEquals(groupBuckets, queueControl.getGroupBuckets());
+//      Assert.assertEquals(groupFirstKey, queueControl.getGroupFirstKey());
+      Assert.assertEquals(lastValue, queueControl.isLastValue());
+//      Assert.assertEquals(lastValueKey, queueControl.getLastValueKey());
```
Comment (michaelpearce-gain) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java]: commented out code......

Comment (michaelpearce-gain) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java]: either comment it in, or remove.

**Comment by michaelpearce-gain** [comment]
Looking better, left a number of comments, one critical bug i found was its possible to raise NPE in QueueImpl constructor, this is because its not following the same pattern as all other fields, lets keep things uniform / follow existing convention.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 23 above)
Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: (nit)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/GroupFirstMessageReference.java (hunk 22 above)
Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/GroupFirstMessageReference.java]: (nit)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 21 above)
Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: (nit)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 20 above)
Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: (nit)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 16 above)
Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: (nit)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 15 above)
Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: (nit)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 14 above)
Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: (nit)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 13 above)
Comment (michaelpearce-gain) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: (nit)

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java (hunk 25 above)
Comment (jbertram) [reply tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/management/ActiveMQServerControlTest.java]: I added these as place-holders to remind me what we don't have implemented yet. I'd like to keep them if possible and then un-comment them when the functionality is implemented on the QueueControl.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 24 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: Fair enough.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/GroupFirstMessageReference.java (hunk 22 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/GroupFirstMessageReference.java]: Fixing...

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 21 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: Will do.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 20 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: Will do.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 17 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: These access changes were left-over from when RingQueue was subclassed. I'll fix them.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java (hunk 13 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: I remember changing this, but I don't remember why. :)  I'll change them back.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/PagedReferenceImpl.java [hunk 26]
```diff
@@ -252,6 +252,16 @@ public void handled() {
       getQueue().referenceHandled(this);
    }
 
+   @Override
+   public void setInDelivery(boolean inDelivery) {
+
```
Comment (wy96f) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/PagedReferenceImpl.java]: This should also be implemented

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 27]
```diff
@@ -974,6 +1018,8 @@ public void addTail(final MessageReference ref) {
    public void addTail(final MessageReference ref, final boolean direct) {
       enterCritical(CRITICAL_PATH_ADD_TAIL);
       try {
+         enforceRing();
```
Comment (wy96f) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: Should put enforceRing below schedule and direct deliver code?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 28]
```diff
@@ -881,13 +919,19 @@ public void addHead(final MessageReference ref, boolean scheduling) {
       enterCritical(CRITICAL_PATH_ADD_HEAD);
       synchronized (this) {
          try {
-            if (!scheduling && scheduledDeliveryHandler.checkAndSchedule(ref, false)) {
-               return;
+            if (ringSize != -1) {
+               enforceRing(ref, scheduling);
```
Comment (wy96f) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: Should put enforceRing below schedule code?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 29]
```diff
@@ -3933,6 +4007,39 @@ void scheduleSlowConsumerReaper(AddressSettings settings) {
...
 
+   private void enforceRing() {
+      if (ringSize != -1) { // better escaping & inlining when ring isn't being used
+         enforceRing(null, false);
+      }
+   }
+
+   private void enforceRing(MessageReference refToAck, boolean scheduling) {
+      if (getMessageCountForRing() >= ringSize) {
+         refToAck = refToAck == null ? messageReferences.poll() : refToAck;
```
Comment (wy96f) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: If a queue is both LVQ and RING queue, when we poll ref from messageReferences, we should also remove it from LastValueQueue::map.

**Comment by wy96f** [comment]
In LastValueQueue::addHead,
https://github.com/apache/activemq-artemis/blob/9d96dda336514d2b49e26e73f2c4c7bd83f6ea31/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/LastValueQueue.java#L152
https://github.com/apache/activemq-artemis/blob/9d96dda336514d2b49e26e73f2c4c7bd83f6ea31/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/LastValueQueue.java#L154,

if ref is replace acked due to over the ring size, we shouldn't put ref in LastValueQueue::map.
